source/simulator/assemblers/stokes.cc

```diff
@@ -31,35 +30,18 @@ namespace aspect
   {
     template <int dim>
     void
-    StokesAssembler<dim>::
-    create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
+    StokesPreconditioner<dim>::
+    execute (internal::Assembly::Scratch::ScratchBase<dim>   &scratch_base,
+             internal::Assembly::CopyData::CopyDataBase<dim> &data_base) const
     {
-      const unsigned int n_points = outputs.viscosities.size();
+      internal::Assembly::Scratch::StokesPreconditioner<dim> &scratch = dynamic_cast<internal::Assembly::Scratch::StokesPreconditioner<dim>& > (scratch_base);
+      internal::Assembly::CopyData::StokesPreconditioner<dim> &data = dynamic_cast<internal::Assembly::CopyData::StokesPreconditioner<dim>& > (data_base);
 
-      if (this->get_parameters().enable_additional_stokes_rhs
-          && outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >() == NULL)
-        {
-          outputs.additional_outputs.push_back(
-            std_cxx11::shared_ptr<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >
-            (new MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> (n_points)));
-        }
-      Assert(!this->get_parameters().enable_additional_stokes_rhs
-             ||
-             outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >()->rhs_u.size()
-             == n_points, ExcInternalError());
-    }
-
-    template <int dim>
-    void
-    StokesAssembler<dim>::
-    preconditioner (const double                                             pressure_scaling,
-                    internal::Assembly::Scratch::StokesPreconditioner<dim>  &scratch,
-                    internal::Assembly::CopyData::StokesPreconditioner<dim> &data) const
-    {
       const Introspection<dim> &introspection = this->introspection();
       const FiniteElement<dim> &fe = this->get_fe();
       const unsigned int stokes_dofs_per_cell = data.local_dof_indices.size();
       const unsigned int n_q_points           = scratch.finite_element_values.n_quadrature_points;
+      const double pressure_scaling = scratch.pressure_scaling;
 
       // First loop over all dofs and find those that are in the Stokes system
       // save the component (pressure and dim velocities) each belongs to.
```

Comment:
I discussed this with @gassmoeller, and this is not something for this pull request, but something that relates to #733 and could be addressed later. To avoid duplicating so much code in the different assemblers, we could move at least the lines below where things in the scratch object are filled out of the assembler, and into the `local_assemble_...` functions. Many of those things are similar for the all Stokes assemblers, or all Stokes preconditioners, and while we would have to fill the things needed by all assemblers in the `local_assemble_...` functions (so potentially also some things that wouldn't be needed later on), it might still not be slower in the case of multiple assemblers being used together, as we would have to go through the loops only once. 
And it would have the advantage that the assemblers would be quite a bit shorter.